### ISSUE TITLE
The result of demo case  between ray tune and flaml is not same.

### ISSUE THREAD

**Issue by zuoxiaojiang** [microsoft/FLAML#195]
I run case studies in page of https://github.com/microsoft/FLAML/tree/main/flaml/tune, which includes the cfo used for flmal.tune and ray.tune. But the result is different between them, so is there something wrong in embding the flmal into ray.tune's framework?

**Comment by qingyun-wu** [comment]
Hi @zuoxiaojiang,

Thank you for your question. 
- Sequential vs parallel: flaml.tune is doing sequential tuning using the provided search algorithm, e.g., cfo, while ray.tune is doing parallel tuning by default. This is the major cause of the difference. 
- Time overhead in ray.tune: when ray.tune is used, there will be some time overhead (in addition to the search time used by the search algorithm) because of initializing and connecting to ray client. So when 'time_budget' is used as a stopping criterion, even if you restrict the number of concurrent trials in ray.tune, we may still see some result difference.

Considering these two differences, if you want to get exactly the same results from flaml.tune and ray.tune, you need to do 1. limit the number of concurrent trials in ray.tune to be 1 using `search_alg = ConcurrencyLimiter(cfo, max_concurrent_trials)` ,  2. use the argument `num_samples`, which is the maximal number of configs to try, instead of `time_budget` (set it to be `None`) as the stopping criterion.

In addition, since there is randomness in both `CFO` and `BlendSearch`, remember to set a random seed before you start the tuning job (for both flaml.tune, and ray.tune), using, e.g., `np.random.seed(100)` .

Please let me know if you need more info or more help. Thank you!

**Comment by zuoxiaojiang** [comment]
Thank you for your&nbsp;reply:&nbsp;



Sorry to reply your email such late, for I'm busy with another work.&nbsp;&nbsp;




I have tried to make a similar result between flaml and ray.tune, as you recommended. But actually, I still can't succeed, my codes are here, and I&nbsp;

need your help to review them in your free time.




This is the code using flaml.tune
from flaml import tune
import numpy as np

np.random.seed(100)


def evaluate_config(config):
    x, y = config["x"], config["y"]
    metric = x + y
    tune.report(metric=metric)


analysis = tune.run(
    evaluate_config,   

    config={
        'x': tune.randint(lower=1, upper=100),
        'y': tune.randint(lower=1, upper=200)
    },
    low_cost_partial_config={},   
    metric='metric',    
    mode='min',        
    num_samples=10,    
    time_budget_s=None,   
    )

print(analysis.best_trial.last_result)  
print(analysis.best_config)



and this is for ray.tune:
from ray.tune.suggest import ConcurrencyLimiter

from ray import tune as raytune
from flaml import CFO, BlendSearch, tune
import time
import numpy as np

np.random.seed(100)

def evaluate_config(config):
    x, y = config["x"], config["y"]
    metric = x + y
    tune.report(metric=metric)

config_search_space = {
        'x': tune.randint(lower=1, upper=100),
        'y': tune.randint(lower=1, upper=200)
    }

# set up CFO
cfo = CFO(low_cost_partial_config={})

search_alg = ConcurrencyLimiter(cfo, 1)

analysis = raytune.run(
    evaluate_config,
    config=config_search_space,
    metric='metric',
    mode='min',
    num_samples=10,
    time_budget_s=None,
    # local_dir='logs/',  # the local directory to store logs
    search_alg=cfo
    )

print(analysis.best_trial.last_result)
print(analysis.best_config)



The following is the results of flamal and ray.tune, the best config are both&nbsp;{'x': 9, 'y': 25}, but the sampled points are not same:









Actually, the above result only occur when random seed set as "100"...




I'm confused for it and I need to use cfo in ray.tune to do my work, so I'm not sure the current version cfo in ray.tune is 100% reliable. So, I'm indeed need your help, thank you again, and expect your next reply~!




Thank you





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/FLAML"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月11日(星期六) 上午9:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [microsoft/FLAML] The result of demo case  between ray tune and flaml is not same. (#195)





 
Hi @zuoxiaojiang,
 
Thank you for your question.
  
Sequential vs parallel: flaml.tune is doing sequential tuning using the provided search algorithm, e.g., cfo, while ray.tune is doing parallel tuning by default. This is the major cause of the difference.
 
Time overhead in ray.tune: when ray.tune is used, there will be some time overhead (in addition to the search time used by the search algorithm) because of initializing and connecting to ray client. So when 'time_budget' is used as a stopping criterion, even if you restrict the number of concurrent trials in ray.tune, we may still see some result difference.
  
Considering these two differences, if you want to get exactly the same results from flaml.tune and ray.tune, you need to do 1. limit the number of concurrent trials in ray.tune to be 1 using search_alg = ConcurrencyLimiter(cfo, max_concurrent_trials) ,  2. use the argument num_samples, which is the maximal number of configs to try, instead of time_budget (set it to be None) as the stopping criterion.
 
In addition, since there is randomness in both CFO and BlendSearch, remember to set a random seed before you start the tuning job (for both flaml.tune, and ray.tune), using, e.g., np.random.seed(100) .
 
Please let me know if you need more info or more help. Thank you!
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by qingyun-wu** [comment]
Hi @zuoxiaojiang, what is the ray tune version you are using?

**Comment by zuoxiaojiang** [comment]
Thank you for the reply. Im using the version of ray.tune 1.5.1&nbsp;


If convenient, can we discuss in WeChat, my WeChat account is ZXJ-5599.&nbsp;


Thank you.&nbsp;


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/FLAML"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月26日(星期天) 上午10:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [microsoft/FLAML] The result of demo case  between ray tune and flaml is not same. (#195)





 
Hi @zuoxiaojiang, what is the ray tune version you are using?
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by qingyun-wu** [comment]
Hi @zuoxiaojiang,

Thank you for providing the information. The required version is [`ray[tune]==1.6.0`](https://github.com/microsoft/FLAML/blob/main/setup.py#L63). In addition, can you please add the step for setting up the searcher before calling flaml.tune (as attached below)?  Let me know if you still get different results.  

And just F.Y.I., we added a consistency test in this file: https://github.com/microsoft/FLAML/blob/main/test/tune/test_flaml_raytune_consistency.py

```
def evaluate_config(config):
    x, y = config["x"], config["y"]
    metric = x + y
    tune.report(metric=metric)


cfo = CFO(low_cost_partial_config={})
analysis = tune.run(
    evaluate_config,

    config={
        'x': tune.randint(lower=1, upper=100),
        'y': tune.randint(lower=1, upper=200)
    },
    low_cost_partial_config={},
    metric='metric',
    mode='min',
    num_samples=10,
    time_budget_s=None,
    search_alg=cfo,
    )

print(analysis.best_trial.last_result)
print(analysis.best_config)

```

**Comment by zuoxiaojiang** [comment]
Thank you for your reply and the case about consistence test, I will try it in some days.

As the case shows, is the "low_cost_partial_config" must be specified, what if unknown?

**Comment by qingyun-wu** [comment]
Hi @zuoxiaojiang,

About `low_cost_partial_config`:

- Definition and purpose: The `low_cost_partial_config ` is a dictionary of subset of the hyperparameter coordinates whose value corresponds to a configuration with known low-cost (i.e., low computation cost for training the corresponding model).  The concept of low/high-cost is meaningful in the case where a subset of the hyperparameters to tune directly affects the computation cost for training the model. For example, `n_estimators` and `max_leaves` are known to affect the training cost of tree-based learners. We call this subset of hyperparameters, *cost-related hyperparameters*. In such scenarios, if you are aware of low-cost configurations for the cost-related hyperparameters, you are recommended to set them as the `low_cost_partial_config`. Using the tree-based method example again, since we know that small `n_estimators` and  `max_leaves` generally correspond to simpler models and thus lower cost, we set `{'n_estimators': 4, 'max_leaves': 4}` as the `low_cost_partial_config` by default (note that `4` is the lower bound of search space for these two hyperparameters), e.g., in [LGBM](https://github.com/microsoft/FLAML/blob/main/flaml/model.py#L215).  Configuring `low_cost_partial_config` helps the search algorithms make more cost-efficient choices.  


- Usage in practice: It is recommended to configure it if there are cost-related hyperparameters in your tuning task and you happen to know the low-cost values for them, but it is not required( It is fine to leave it the default value, i.e., `None`).

- How does it work: `low_cost_partial_config` if configured, will be used as an initial point of the search. It also affects the search trajectory. For more details about how does it play a role in the search algorithms, please refer to the papers about the search algorithms used: Section 2 of [Frugal Optimization for Cost-related Hyperparameters (CFO)](https://arxiv.org/pdf/2005.01571.pdf) and Section 3 of [Economical Hyperparameter Optimization with Blended Search Strategy (BlendSearch)](https://openreview.net/pdf?id=VbLH04pRA3). 

Please let us know if you need more help. Thank you!

**Comment by sonichi** [comment]
@qingyun-wu and @zuoxiaojiang is this issue resolved?

**Comment by zuoxiaojiang** [comment]
> @qingyun-wu and @zuoxiaojiang is this issue resolved?



> Hi @zuoxiaojiang, in addition to updating ray version, did you use the code I provided?



> Hi @zuoxiaojiang, in addition to updating ray version, did you use the code I provided?

Hi @qingyun-wu, today I try the code for testing consistency which you provided, and finally get a correct and same results. Thank you for your patiente help. 

I found my code use the library which is `from flaml import CFO`, while you use the `from flaml.searcher.blendsearch import CFO`. I think this is the course for difference.

**Comment by qingyun-wu** [comment]
> > Hi @zuoxiaojiang, in addition to updating ray version, did you use the code I provided?
> 
> Hi @qingyun-wu, today I try the code for testing consistency which you provided, and finally get a correct and same results. Thank you for your patiente help.
> 
> I found my code use the library which is `from flaml import CFO`, while you use the from `flaml.searcher.blendsearch import CFO`. I think this is the cause for difference.

Great! Thanks for the follow-up. I will close this issue if you don't have further questions.
